File: algorithm/GO/两个文件的相对路径.go
package main

import (
	"strings"
	"fmt"
	"github.com/pkg/errors"
)

func main() {
	fmt.Println(getRelativePath(
		"/Users/alvin/project/go/src/colly/main.go",
		"/Users/alvin/project/collect/algorithm/GO/插入排序.go"))
}

func getRelativePath(path1, path2 string) (string, error) {
	if path1 == "" || path2 == "" {
		return "", errors.New("绝对路径不可以为空")
	}

	arr1 := strings.Split(path1[1:], "/")
	arr2 := strings.Split(path2[1:], "/")
	depth := 0
	for i := 0; i < len(arr1) && i < len(arr2); i++ {
		if arr1[i] == arr2[i] {
			depth ++
		} else {
			break
		}
	}
	prefix := ""
	if len(arr1)-depth-1 <= 0 {
		prefix = "./"
	} else {
		for i := len(arr1) - depth - 1; i > 0; i -- {
			prefix += "../"
		}
	}
	fmt.Println(depth)
	if len(arr2)-depth > 0 {
		prefix += strings.Join(arr2[depth:], "/")
	}
	return prefix, nil
}

File: algorithm/两个文件的相对路径.php
<?php

/**
 * 两个文件的相对路径
 *
 * @param $path1:路径一
 * @param $path2:路径二
 * @return string
 */
function getRelativePath($path1, $path2)
{
    $arr1 = explode('/', dirname($path1));
    $arr2 = explode('/', dirname($path2));
    for ($i = 0, $len = count($arr2); $i < $len; $i++) {
        if ($arr1[$i] != $arr2[$i]) {
            break;
        }
    }

    if ($i < $len) {
        $return_path = array_fill(0, $len - $i, '..');
    }
    //$b相对于$a
    $return_path = array_merge($return_path, array_slice($arr1, $i));
    //$a相对于$b
    // $return_path=array_merge($return_path,array_slice($arr2,$i));
    return implode('/', $return_path);

}

var_dump(getRelativePath("/Users/alvin/project/markdown/事务.md",
    "/Users/alvin/project/collect/algorithm/两个文件的相对路径.php"));